### PR TITLE
Feature/python3

### PR DESCRIPTION
**Pull request by guiguem**
Adapt the T2K-DM tools for python3. I used 2to3 for adaptation (fixed the relative path issues with __ini__.py).
Removing toml which seems not to be used by setup.py (and cause big issues with python3)

### CODE REVIEW

File: t2kdm/interactive.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def ls(remotepath, *args, **kwargs):
         entries = dm.iter_ls(remotepath, *args, **kwargs)
     else:
         entries = dm.iter_ls_se(remotepath, *args, se=se, **kwargs)
-    if long:
+    if int:
```
Comment (ast0815) [inline t2kdm/interactive.py:159]: This is defintely a wrong change. `long` is a variable name.

File: t2kdm/configuration.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def __init__(self, filename=None, defaults=default_values):
         """
 
         # Load the default values
-        for key, val in defaults.items():
+        for key, val in list(defaults.items()):
```
Comment (ast0815) [inline t2kdm/configuration.py:58]: This `list` is not needed, since we directly iterate over the return value of `items`.

File: t2kdm/cache.py [hunk 3]
```diff
@@ -27,13 +27,13 @@ def __init__(self, cache_time=60):
...
         """Remove old entries from the cache."""
-        for key in self.cache.keys():
+        for key in list(self.cache.keys()):
             if not self.cache[key].is_valid():
                 del self.cache[key]
 
     def flush(self):
         """Remove all entries from the cache."""
-        for key in self.cache.keys():
+        for key in list(self.cache.keys()):
```
Comment (ast0815) [inline t2kdm/cache.py:36]: This is not needed, since we iterate directly.

File: t2kdm/cache.py [hunk 4]
```diff
@@ -27,13 +27,13 @@ def __init__(self, cache_time=60):
 
     def clean(self):
         """Remove old entries from the cache."""
-        for key in self.cache.keys():
+        for key in list(self.cache.keys()):
```
Comment (ast0815) [inline t2kdm/cache.py:30]: This is not needed since we iterate directly.

File: t2kdm/backends.py [hunk 5]
```diff
@@ -1081,7 +1081,7 @@ def _remove(self, surl, lurl, last=False, verbose=False, **kwargs):
         if not ret["OK"]:
             raise BackendException("Failed: %s" % (ret["Message"]))
 
-        for lurl, error in ret["Value"]["Failed"].items():
+        for lurl, error in list(ret["Value"]["Failed"].items()):
```
Comment (ast0815) [inline t2kdm/backends.py:1084]: This `list` is not needed, since we iterate directly over the output of `items`.

File: t2kdm/backends.py [hunk 6]
```diff
@@ -815,7 +815,7 @@ def _iter_directory(self, lurl):
         ret = self.fc.listDirectory(lurl)
         if not ret["OK"]:
             raise BackendException("Failed to list path '%s': %s", lurl, ret["Message"])
-        for path, error in ret["Value"]["Failed"].items():
+        for path, error in list(ret["Value"]["Failed"].items()):
```
Comment (ast0815) [inline t2kdm/backends.py:818]: Not needed, since we iterate directly.

File: t2kdm/backends.py [hunk 7]
```diff
@@ -790,11 +790,11 @@ def _get_dir_entry(self, lurl, infodict=None):
                 raise BackendException(
                     "Failed to list path '%s': %s", lurl, md["Message"]
                 )
-            for path, error in md["Value"]["Failed"].items():
+            for path, error in list(md["Value"]["Failed"].items()):
                 if "No such file" in error:
                     # File does not exist, maybe a directory?
                     md = self.fc.getDirectoryMetadata(lurl)
-                    for path, error in md["Value"]["Failed"].items():
+                    for path, error in list(md["Value"]["Failed"].items()):
```
Comment (ast0815) [inline t2kdm/backends.py:797]: Not needed, since we iterate directly.

File: t2kdm/backends.py [hunk 8]
```diff
@@ -790,11 +790,11 @@ def _get_dir_entry(self, lurl, infodict=None):
                 raise BackendException(
                     "Failed to list path '%s': %s", lurl, md["Message"]
                 )
-            for path, error in md["Value"]["Failed"].items():
+            for path, error in list(md["Value"]["Failed"].items()):
```
Comment (ast0815) [inline t2kdm/backends.py:793]: Not needed, since we iterate directly.

File: t2kdm/backends.py [hunk 9]
```diff
@@ -763,7 +763,7 @@ def __init__(self, **kwargs):
     def _check_return_value(ret):
         if not ret["OK"]:
             raise BackendException("Failed: %s", ret["Message"])
-        for path, error in ret["Value"]["Failed"].items():
+        for path, error in list(ret["Value"]["Failed"].items()):
```
Comment (ast0815) [inline t2kdm/backends.py:766]: Not needed, since we iterate directly.

**Review by ast0815 — COMMENTED** [review]
Seems like `2to3` was a bit eager and changed some stuff that does not need changing. Does this version still work with Python 2.7 and the older Dirac setups?

File: t2kdm/interactive.py (hunk 1 above)
Comment (ast0815) [reply t2kdm/interactive.py:159]: If `long` is also a variable type in Python, maybe we should rename the variable.

Comment (guiguem) [reply t2kdm/interactive.py:159]: It seems it is for python3... 😅

Comment (guiguem) [reply t2kdm/interactive.py]: Changed it to long_str

File: t2kdm/backends.py [hunk 10]
```diff
@@ -885,13 +884,13 @@ def _replicas(self, lurl, **kwargs):
...
 
-        return rep.values()
+        return list(rep.values())
 
     def _exists(self, surl, **kwargs):
         try:
             ret = self._ls_se_cmd(surl, "-d", "-l", **kwargs).strip()
         except sh.ErrorReturnCode as e:
-            if "No such file" in e.stderr:
+            if "No such file" in str(e):
```
Comment (guiguem) [inline t2kdm/backends.py:893]: With my setup, I cannot get stderr from e; only a long string which I just print for debugging purpose: not ideal...

Comment (ast0815) [reply t2kdm/backends.py:893]: What version of the python `sh` package do you have?

Comment (guiguem) [reply t2kdm/backends.py:893]: @ast0815 I am using sh version 1.14.2 in the container

Comment (guiguem) [reply t2kdm/backends.py:893]: Pypi says this is the latest version

**Comment by ast0815** [comment]
Any news about this?

**Comment by ast0815** [comment]
Implemented this in #62.

**Pull request closed without merging**